### DISCUSSION TITLE
Saving of resized animated GIF

### DISCUSSION THREAD

**Discussion by Jacherr** [libvips/libvips#3962] [Q&A] [answered]
Hi, it's me again :P. I'm attempting to load an animated GIF, resize all of its frames via `vips_resize` and then save it again, as an animated GIF.

I have the following code:
```c
        VipsImage* image = vips_image_new_from_buffer(input_image, input_length, "", "n", -1, NULL);

        int page_width = vips_image_get_width(image);
        int page_height = vips_image_get_page_height(image);
        int n_pages = vips_image_get_n_pages(image);
        VipsImage* pages[n_pages];

        for(int i = 0; i < n_pages; i++) {
            vips_crop(image, &pages[i], 0, i * page_height, page_width, page_height, NULL);
            vips_resize(pages[i], &pages[i], scale, NULL), EXEC_OPERATION_VIPS_ERR);
        }

        VipsImage* out;

        vips_arrayjoin(pages, &out, n_pages, "across", 1, NULL);
        vips_gifsave_buffer(out, (void**) output_image, (size_t*) output_length, NULL);
```

However, attempting to save this GIF produces all kinds of strange effects depending on the `scale` provided. Is this due to the fact that `vips_resize` does not change the image resolution (as per the documentation)? If so, what's the best way to correct this? I wasn't able to find much in the documentation.

**Comment by jcupitt** [comment, marked as answer]
Hi again, you need to set the new `page_height` so the gif saver knows how to write each page of the output image.

You don't need to chop into pages and then reassemble -- just pick a scale factor which will get the page boundaries to land in the right place. There's some sample code in the `thumbnail` sources:

https://github.com/libvips/libvips/blob/master/libvips/resample/thumbnail.c#L782-L791

But I'd just call `thumbnail` instead. It'll handle things like premultiply for you too.

```C
VipsImage *thumb = vips_thumbnail_buffer(input_data, input_length, target_width,
    "option_string", "n=-1",
    NULL);
void *data;
size_t length;
vips_gifsave_buffer(thumb, &data, &length, NULL);
```

**Reply by Jacherr** [reply]
Ah yep, using thumbnail worked great. As an aside, this seems to produce an odd line of pixels at the top of the output. My input is this:
![attachment](https://github.com/libvips/libvips/assets/40544507/ed01b3d1-2d73-489f-8b0d-b9ee2aed8517)

And I'm running the following code:
```c
        int target_width = (int) ((double) vips_image_get_width(image) * scale);
        int target_height = (int) ((double) vips_image_get_height(image) * scale);
        VipsImage * out;

        vips_thumbnail_image(image, &out, target_width, "height", target_height, NULL);
        vips_gifsave_buffer(out, (void**) output_image, (size_t*) output_length, NULL);
```

(I'm using `vips_thumbnail_image` here as it works better with my other code)

Where the output is as follows:
![attachment](https://github.com/libvips/libvips/assets/40544507/ba77a051-29fa-45b0-8e65-c8497167c118)

The top line of pixels seems to be a bit off, is this an artifact of the resize method being used or potentially a quirk of GIF encoding?

**Reply by jcupitt** [reply]
Yes, that's the resize method interacting with your specific image. It might be possible to improve it, I'll have a look.

In the meantime, you'll have to split it to separate frames, resize, then rejoin :(

**Reply by jcupitt** [reply]
I wouldn't use `thumbnail_image` -- it can be very slow. It's only there for emergencies.

**Reply by Jacherr** [reply]
Ah, alright - I'll take the splitting approach for now, then. It works well enough for my use case!

For setting the page height - is it just a matter of doing `vips_image_set_int(out, "page_height", vips_image_get_page_height(in))`? And in addition - is it `page_height` or `page-height`? The documentation seems to allude to the latter, but just making sure.

**Reply by jcupitt** [reply]
It'll copy over the old page height automatically. You need to set the *new* page height, ie. the height of the resized frames.

You can use `-` or `_`.

**Reply by Jacherr** [reply]
Great, this works perfectly. Thank you very much again.
